File: backend/app/models/sync.py
"""Pydantic-модели синхронизации прогресса (user_answers + fsrs_state).

Клиент - master of FSRS math (считает в Web Worker). Backend хранит журнал
ответов и зеркало FSRS-состояния для синхронизации между устройствами.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class EntityType(StrEnum):
    CARD = "card"
    ACCREDITATION_QUESTION = "accreditation_question"


class AnswerSource(StrEnum):
    FEED = "feed"
    PREP = "prep"
    EXAM = "exam"
    BROWSE = "browse"


class FsrsSource(StrEnum):
    FEED = "feed"
    PREP = "prep"


class AnswerRecord(BaseModel):
    entity_type: EntityType
    entity_id: str = Field(..., min_length=1, max_length=128)
    is_correct: bool
    answered_at_ms: int = Field(..., gt=0)
    time_spent_ms: int | None = Field(default=None, ge=0)
    source: AnswerSource = AnswerSource.FEED
    idempotency_key: str = Field(..., min_length=1, max_length=128)


class FsrsStateDelta(BaseModel):
    entity_id: str = Field(..., min_length=1, max_length=128)
    source: FsrsSource = FsrsSource.FEED
    state: dict[str, Any] = Field(
        ..., description="FSRS-состояние как JSON: stability, difficulty, due, reps, lapses, ..."
    )
    updated_at_ms: int = Field(..., gt=0)


class BatchAnswersRequest(BaseModel):
    answers: list[AnswerRecord] = Field(default_factory=list, max_length=500)
    fsrs_updates: list[FsrsStateDelta] = Field(default_factory=list, max_length=500)


class BatchAnswersResponse(BaseModel):
    answers_accepted: int
    answers_duplicates: int
    fsrs_upserts: int


class FsrsStateRow(BaseModel):
    entity_id: str
    source: FsrsSource
    state: dict[str, Any]
    updated_at_ms: int


class FsrsStateResponse(BaseModel):
    rows: list[FsrsStateRow]
    total: int


class AnswerRow(BaseModel):
    entity_type: EntityType
    entity_id: str
    is_correct: bool
    answered_at_ms: int
    time_spent_ms: int | None
    source: AnswerSource


class AnswersSinceResponse(BaseModel):
    rows: list[AnswerRow]
    total: int

File: src/lib/backend/sync.ts
/**
 * Клиент progress-sync API. Контракт 1:1 с `backend/app/models/sync.py`.
 */

import { backendFetch } from "@/lib/backendClient";

export type EntityType = "card" | "accreditation_question";
export type AnswerSource = "feed" | "prep" | "exam" | "browse";
export type FsrsSource = "feed" | "prep";

export interface AnswerRecord {
  entity_type: EntityType;
  entity_id: string;
  is_correct: boolean;
  answered_at_ms: number;
  time_spent_ms?: number;
  source: AnswerSource;
  idempotency_key: string;
}

export interface FsrsStateDelta {
  entity_id: string;
  source: FsrsSource;
  state: Record<string, unknown>;
  updated_at_ms: number;
}

export interface BatchAnswersRequest {
  answers: AnswerRecord[];
  fsrs_updates: FsrsStateDelta[];
}

export interface BatchAnswersResponse {
  answers_accepted: number;
  answers_duplicates: number;
  fsrs_upserts: number;
}

export interface FsrsStateRow {
  entity_id: string;
  source: FsrsSource;
  state: Record<string, unknown>;
  updated_at_ms: number;
}

export interface FsrsStateResponse {
  rows: FsrsStateRow[];
  total: number;
}

export interface AnswerRow {
  entity_type: EntityType;
  entity_id: string;
  is_correct: boolean;
  answered_at_ms: number;
  time_spent_ms: number | null;
  source: AnswerSource;
}

export interface AnswersSinceResponse {
  rows: AnswerRow[];
  total: number;
}

export function submitBatch(
  payload: BatchAnswersRequest,
  opts?: { signal?: AbortSignal },
): Promise<BatchAnswersResponse> {
  return backendFetch<BatchAnswersResponse>("/api/answers/batch", {
    method: "POST",
    body: payload,
    signal: opts?.signal,
  });
}

export function fetchFsrsState(
  query: { since?: number; source?: FsrsSource } = {},
  opts?: { signal?: AbortSignal },
): Promise<FsrsStateResponse> {
  return backendFetch<FsrsStateResponse>("/api/fsrs/state", {
    query,
    signal: opts?.signal,
  });
}

export function fetchAnswersSince(
  query: { since?: number; limit?: number } = {},
  opts?: { signal?: AbortSignal },
): Promise<AnswersSinceResponse> {
  return backendFetch<AnswersSinceResponse>("/api/answers/since", {
    query,
    signal: opts?.signal,
  });
}
